lua/lib/lfospec.lua

```diff
@@ -0,0 +1,419 @@
+-- LFOs for general-purpose scripting
+-- @module lib.lfospec
+-- inspired by contributions from @markwheeler (changes), @justmat (hnds), and @sixolet (toolkit)
+-- added by @dndrks
+
+local LFO = {}
+LFO.__index = LFO
+
+local lfo_rates = {1/16,1/8,1/4,5/16,1/3,3/8,1/2,3/4,1,1.5,2,3,4,6,8,16,32,64,128,256,512,1024}
+local lfo_rates_as_strings = {"1/16","1/8","1/4","5/16","1/3","3/8","1/2","3/4","1","1.5","2","3","4","6","8","16","32","64","128","256","512","1024"}
+
+local update_freq = 96
```

Comment:
If update_freq is constant, it might make sense (like lattice) to have all LFOs share a clock. Or... use a lattice?

Separately, it might make sense to have update_freq settable. I know that 96ppqn can be a lot if the LFO is triggering a lot on each step.

Reply:
ah yeah, both points make a lot of sense -- thank you! using a lattice to manage the complexity of having groups of LFOs share a clock in the library itself (which the first draft of this that was reverted did, but at the expense of scripting legibility) would be good

Reply:
Yeah, I can probably give it a try later today. Shouldn't be that much of a change. There's some options here, tho, the design space is pretty big:
* Reintroduce the concept of an LFO group, LFO groups share a clock/lattice and a ppqn?
* Leave LFOs on separate clocks, each one picks its ppqn?
* One global ppqn, but settable by the scripter, all LFOs share the same clock?
* One global "base" ppqn, all LFOs on the same clock, (96 ppqn, perhaps not even settable) but each LFO can pick a "bang divider" which allows it to update less frequently — you might pick 2 for a 48ppqn option, or even something like 96 for an LFO that has a period of several minutes.
* ~~LFOs will pick a "bang divider" depending on their frequency, aiming for a resolution that makes sense for them?~~ (This one seems like it'd be a little hard to pick a "goldilocks" value for every context, maybe let's not pick this) 

Which seems to fit best? I think I gently favor the "bang divider" option in that you can just use the lib and it'll bang at 96 like the current PR, but if you find it is overloading your system you can slow its roll pretty easily. And I don't imagine many situations where you'd want to slow its roll by a factor of 1.5 or whatever, slowing the roll by a factor of 2 or 3 or 4 makes a lot more sense. My one worry is what to call the "bang divider" argument, "bang divider" isn't immediately obvs what it means.

Reply:
love your solution of leveraging lattice and using `:set_division` to manage unique ppqn's per LFO -- and this way, just consumes one clock!